src/home_robot/home_robot/mapping/voxel.py

```diff
@@ -492,18 +552,65 @@ def _show_pytorch3d(self, **plot_scene_kwargs):
         )
         return fig
 
+    def sample_explored(self, robot: Optional[Robot] = None) -> Optional[np.ndarray]:
+        """Return obstacle-free xy point in explored space"""
+        obstacles, explored = self.get_2d_map()
+        valid_indices = np.argwhere(
+            np.bitwise_and(
+                np.bitwise_not(obstacles.cpu().numpy()), explored.cpu().numpy()
+            )
+            == 1
+        )
+        if len(valid_indices) > 0:
+            random_index = np.random.choice(len(valid_indices))
+            return self.grid_coords_to_xy(valid_indices[random_index])
+        else:
+            return None
```

Comment:
This function can probably be written in torch and then cast to numpy at the last second

Reply:
fixed